### PR TITLE
Add guiBlur and guiFocus functions (Mantis #7448)

### LINKED ISSUE #457: Add function guiEditSetActive() or similar
**Is your feature request related to a problem? Please describe.**
Function to activate gui Edit or Memo, which puts the cursor inside automaticly, so player can concentrate on typing, without having to do it slow, manually.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**
https://github.com/multitheftauto/mtasa-blue/pull/365
https://bugs.mtasa.com/view.php?id=7448

Someone told me this was possible with using guiBringToFront and guiEditSetCaretIndex.

Just saw this issue, haven't tried it yet.

Confirmed that it works like:
http://pastebin.com/mxQme0td

It also works without, guiEditSetCaretIndex( edit, 1 ), depending on the rest of the gui.

### PR DESCRIPTION
**Pull request by patrikjuvonen**
### Mantis Bug Tracker issue
[7448](https://bugs.multitheftauto.com/view.php?id=7448), fixes #457

### Summary
- Add guiBlur function, which "de-focuses" the specified gui element.
`bool guiBlur ( element guiElement )`
- Add guiFocus function, which "focuses" the specified gui element.
~~`bool guiFocus ( element guiElement [, bool bMoveCaretToEnd = true ] )`~~
   `bool guiFocus ( element guiElement )`
- [onClientGUIBlur](https://wiki.multitheftauto.com/wiki/OnClientGUIBlur) and [onClientGUIFocus](https://wiki.multitheftauto.com/wiki/OnClientGUIFocus) events work with these functions.
- Current approach (or hack, rather) is to use [guiBringToFront](https://wiki.multitheftauto.com/wiki/GuiBringToFront) to do a "focus", and [guiMoveToBack](https://wiki.multitheftauto.com/wiki/GuiMoveToBack) to do a "blur", which is wrong by looking at the naming of those functions and what they're *actually* supposed to do...
- Should work on any gui element.

### Requesting scripting feedback
The `bMoveCaretToEnd` argument is used only on editboxes and memos to move the caret to the end of the text content. I introduced this argument so it would be easier for things like username/password fields where you usually want to autofocus to the end of the input field instead of the default CEGUI behavior where it focuses always to beginning of the field. With this approach to set a specific caret index, you'd have to use use [guiEditSetCaretIndex](https://wiki.multitheftauto.com/wiki/GuiEditSetCaretIndex) or [guiMemoSetCaretIndex](https://wiki.multitheftauto.com/wiki/GuiMemoSetCaretIndex) separate from this function.

Another approach would be to have a `uiCaretIndex` argument instead of `bMoveCaretToEnd` for more specific caret focusing without having to use [guiEditSetCaretIndex](https://wiki.multitheftauto.com/wiki/GuiEditSetCaretIndex) or [guiMemoSetCaretIndex](https://wiki.multitheftauto.com/wiki/GuiMemoSetCaretIndex) when focusing. This means scripters would have to manually input the length of the text to set the caret to the end.

I'm requesting feedback from you to vote on which is better.

### CODE REVIEW

**Comment by qaisjp** [comment]
Regarding `Requesting scripting feedback` - I think we should do what browsers do, and set it to the most recent caret position.

**Comment by patrikjuvonen** [comment]
> Regarding `Requesting scripting feedback` - I think we should do what browsers do, and set it to the most recent caret position.

@qaisjp Thanks for the feedback, I agree. I've now changed the behavior so that by default the caret position is set to the previous caret position. I still kept the `bMoveCaretToEnd` argument though, makes script code cleaner for those who want that behavior.

**Review by qaisjp — APPROVED** [review]
> I still kept the bMoveCaretToEnd argument though, makes script code cleaner for those who want that behavior.

This pull request looks ready to merge! Just not sure on the extra arg — it seems like it's just to save an extra line, which feels unnecessary to me.

@multitheftauto/blue-maintainers, what are your thoughts, can you weigh in on the extra argument? Please send another review and merge!

**Comment by botder** [comment]
I don't mind the extra argument.

**Comment by qaisjp** [comment]
> Thanks for the feedback, I agree. I've now changed the behavior so that by default the caret position is set to the previous caret position. I still kept the `bMoveCaretToEnd` argument though, makes script code cleaner for those who want that behavior.

But then server owners will have to keep track of whether the player moved the caret, and will always set `bMoveCaretToEnd=true`... right?

This is what I am imagining:

- There's a login panel with username and password
- User types in username (this caret positioning is now set to end of this field)
- User types in password (this caret positioning is now set to end of this field)
- User clicks submit (focus taken off password field)
- Password is incorrect
- Password field is focused to, which moves caret to the last caret position, which is the end of that field

Which is as intended, without an extra line or extra parameter... right?

**Comment by patrikjuvonen** [comment]
Yes. However add a few complications to your image such as typing an error in the middle of your username and clicking at that specific caret position to fix it. Now your caret position is not in the end anymore. I don't know what the big hassle is about.

**Comment by qaisjp** [comment]
Indeed, I thought about that. So I checked what browsers do, and they don't set the caret position to the end. I don't think we should either!

> hassle

I'm trying to avoid adding extra cruft to the Lua API when an extra line of code works just fine.

**Comment by patrikjuvonen** [comment]
I understand that we don't want to add more redundant stuff. In this case the behavior is useful and reasonable. We're not providing the argument for setting the caret position to a specified index but rather just to the end. It's more of a choice/preference parameter and I believe it makes sense and doesn't haunt me at night thinking we have an extra argument in our code :)

File: Client/mods/deathmatch/logic/luadefs/CLuaGUIDefs.cpp
Comment (botder) [inline Client/mods/deathmatch/logic/luadefs/CLuaGUIDefs.cpp]: You can reduce the two lines into `lua_pushboolean(luaVM, CStaticFunctionDefinitions::GUIBlur(*guiElement));`

File: Client/mods/deathmatch/logic/luadefs/CLuaGUIDefs.cpp
Comment (botder) [inline Client/mods/deathmatch/logic/luadefs/CLuaGUIDefs.cpp:1920]: Change this line to `lua_pushnil(luaVM)`. Bad arguments must result in **nil**.

File: Client/mods/deathmatch/logic/luadefs/CLuaGUIDefs.cpp
Comment (botder) [inline Client/mods/deathmatch/logic/luadefs/CLuaGUIDefs.cpp:1940]: Change this line to `lua_pushnil(luaVM)`. Bad arguments must result in **nil**.

File: Client/mods/deathmatch/logic/luadefs/CLuaGUIDefs.cpp
Comment (botder) [inline Client/mods/deathmatch/logic/luadefs/CLuaGUIDefs.cpp]: Use the return value of `CStaticFunctionDefinitions::GUIFocus` instead of `true`

File: Client/mods/deathmatch/logic/luadefs/CLuaGUIDefs.cpp
Comment (botder) [inline Client/mods/deathmatch/logic/luadefs/CLuaGUIDefs.cpp]: This block will execute even if `CStaticFunctionDefinitions::GUIFocus` returns false, please correct.

**Review by botder — DISMISSED** [review]

**Comment by patrikjuvonen** [comment]
@botder Thanks for your review. I've updated the code accordingly.

**Review by qaisjp — CHANGES_REQUESTED** [review]
I am pretty sure about the additional `bMoveCaretToEnd` being unnecessary. (Neither does JavaScript have this sort of argument for focus)

Also, I think @botder's review was incorrect. guiBlur and guiFocus both only return `true` for success, and so should return `false` for any sort of failure. `nil` is only returned for getters.

**Comment by patrikjuvonen** [comment]
> I am pretty sure about the additional `bMoveCaretToEnd` being unnecessary. (Neither does JavaScript have this sort of argument for focus)
> 
> Also, I think @botder's review was incorrect. guiBlur and guiFocus both only return `true` for success, and so should return `false` for any sort of failure. `nil` is only returned for getters.

Thanks for the second review. I've now adjusted the functions accordingly.

**Review by qaisjp — APPROVED** [review]

**Pull request merged**
